File: ztalloc/ztalloc.py
import sys
import queue

def solve(Lin, Rin, Lout, Rout):
	seen_set = {(Lin, Rin)}
	q = queue.Queue()
	q.put((Lin,Rin,'i'))
	seen_states = 0	
	if Lin >= Lout and Rin <= Rout:
		return "EMPTY"
	
	while(not q.empty()):
		seen_states = seen_states + 1

		if seen_states > 1000000:
			break		
			
		(a, b, c) = q.get()
		a2, b2 = a//2, b//2
		a3, b3 = 3*a+1, 3*b+1	
	
		if (a2, b2) not in seen_set:
			seen_set.add((a2,b2))
			s = c + 'h'
			q.put((a2,b2,s))
			if a2 >= Lout and b2 <= Rout:
				return s[1:]
				
		if (a3, b3) not in seen_set and b3 < 1000000:
			seen_set.add((a3,b3))
			s = c + 't'
			q.put((a3,b3,s))	
			if a3 >= Lout and b3 <= Rout:
				return s[1:]

	return "IMPOSSIBLE"
				
with open(sys.argv[1],"r") as f:
	Q = int(next(f))
	#print(Q)	
	for line in f:	
		Lin, Rin, Lout, Rout = map(int, line.split())
		solution = solve(Lin, Rin, Lout, Rout)
		print(solution)
		#print(Lin, Rin, Lout, Rout)

File: Ztalloc.java
import java.util.Scanner; 
import java.io.File;

public class Ztalloc {
  // The main function.
  public static void main(String args[]) throws Exception {
	File file = new File(args[0]);
	Scanner scanner = new Scanner(file);
	int Q = scanner.nextInt();
	while(Q-- > 0) {
		int Lin = scanner.nextInt();
		int Rin = scanner.nextInt();
		int Lout = scanner.nextInt();
		int Rout = scanner.nextInt();
		  
		if (Lin >= Lout && Rin <= Rout) {
	    	System.out.print("EMPTY");
	    }
	    else {
			Solver solver = new BFSolver();
		    State initial = new ZtallocState(Lin, Rin, 'i', Lout, Rout, null);
		    State result = solver.solve(initial);
		    if (result == null) {
		      System.out.print("IMPOSSIBLE");
		    } 
		    else {
		      printSolution(result);
		    }
		 }
		System.out.println("");
	}
	scanner.close();
  }

  // A recursive function to print the states from the initial to the final.
  private static void printSolution(State s) {
    if (s.getPrevious() != null) {
      printSolution(s.getPrevious());
    }
    System.out.print(s);
  }
}